xds/src/main/java/io/grpc/xds/XdsServerWrapper.java

```diff
@@ -430,15 +526,226 @@ private void handleConfigNotFound(StatusException exception) {
+            }
+          }
+        });
+      }
+    }
+
+    private boolean isPermanentError(Status error) {
+      return EnumSet.of(
+              Status.Code.INTERNAL,
+              Status.Code.INVALID_ARGUMENT,
+              Status.Code.FAILED_PRECONDITION,
+              Status.Code.PERMISSION_DENIED,
+              Status.Code.UNAUTHENTICATED)
+              .contains(error.getCode());
+    }
+  }
+
+  @VisibleForTesting
+  final class ConfigApplyingInterceptor implements ServerInterceptor {
+    private final ServerInterceptor noopInterceptor = new ServerInterceptor() {
+      @Override
+      public <ReqT, RespT> Listener<ReqT> interceptCall(ServerCall<ReqT, RespT> call,
+          Metadata headers, ServerCallHandler<ReqT, RespT> next) {
+        return next.startCall(call, headers);
+      }
+    };
+
+    @Override
+    public <ReqT, RespT> Listener<ReqT> interceptCall(ServerCall<ReqT, RespT> call,
+        Metadata headers, ServerCallHandler<ReqT, RespT> next) {
+      ServerRoutingConfig routingConfig = call.getAttributes().get(ATTR_SERVER_ROUTING_CONFIG);
+      if (routingConfig == null) {
+        call.close(
+                Status.UNAVAILABLE.withDescription("Missing xDS routing config"),
+                new Metadata());
+        return new Listener<ReqT>() {};
+      }
+      VirtualHost virtualHost = RoutingUtils.findVirtualHostForHostName(
+          routingConfig.virtualHosts(), call.getAuthority());
+      if (virtualHost == null) {
+        call.close(
+                Status.UNAVAILABLE.withDescription("Could not find xDS virtual host matching RPC"),
+                new Metadata());
+        return new Listener<ReqT>() {};
+      }
+      Route selectedRoute = null;
+      Map<String, FilterConfig> selectedOverrideConfigs =
+              new HashMap<>(virtualHost.filterConfigOverrides());
+      MethodDescriptor<ReqT, RespT> method = call.getMethodDescriptor();
+      for (Route route : virtualHost.routes()) {
+        if (RoutingUtils.matchRoute(
+            route.routeMatch(), "/" + method.getFullMethodName(), headers, random)) {
+          selectedRoute = route;
+          selectedOverrideConfigs.putAll(route.filterConfigOverrides());
+          break;
+        }
+      }
+      if (selectedRoute == null) {
+        call.close(
+            Status.UNAVAILABLE.withDescription("Could not find xDS route matching RPC"),
+            new Metadata());
+        return new ServerCall.Listener<ReqT>() {};
+      }
+      List<ServerInterceptor> filterInterceptors = new ArrayList<>();
+      for (NamedFilterConfig namedFilterConfig : routingConfig.httpFilterConfigs()) {
+        FilterConfig filterConfig = namedFilterConfig.filterConfig;
+        Filter filter = filterRegistry.get(filterConfig.typeUrl());
+        if (filter instanceof ServerInterceptorBuilder) {
+          ServerInterceptor interceptor =
+              ((ServerInterceptorBuilder) filter).buildServerInterceptor(
```

Comment:
This could be fine initially, but we want to build the interceptors up-front and reuse them. We want most processing in the control plane that manages updates and then have this data plane do as little as possible.

Reply:
Im not sure pre-build the interceptor is possible, mainly because the override config is different per each request.(I assume that the interceptor is immutable per each config:configoverride data )Unless if we guess that override is not used often and only cover the most frequently used cases.

Reply:
We'd create one set of interceptors per Route/WeightedCluster.